casper/src/main/resources/AsyncMap4.rho

```diff
@@ -0,0 +1,213 @@
+              }
+            }
+            "setBit" => {
+              match args {
+                [bit, ret] => {
+                  @ret!(Nil) |
+                  MakeCell!(
+                    // val | (1 << bit)
+                    // Bitwise operators would be really nice to have!
+                    (val % powers.nth(bit)) +
+                      (val / powers.nth(bit + 1)) * powers.nth(bit + 1) +
+                      powers.nth(bit),
+                    *cell
+                  )
+                }
+                x => { o!(["ERROR: MATCH FAIL. MakeCell.get line 41", x]) }
+              }
+            }
+            "die" => { Nil }
+            x => { o!(["ERROR: MATCH FAIL. MakeCell line 45", x]) }
+          }
+        }
+      } |
+
+      contract ByteArrayToNybbleList(@ba, @n, @len, @acc, ret) = {
+        if (n == len) {
+          ret!(acc)
+        } else {
+          ByteArrayToNybbleList!(ba, n+1, len, acc ++ [ ba.nth(n) % 16, ba.nth(n) / 16 ], *ret)
+        }
+      } |
+
+      contract AsyncMap(@"init", ret) = {
+        new map in {
+          MakeCell!(0, [*map, []]) |
+          ret!(*map)
+        }
+      } |
+
+      contract AsyncMapGetter(@map, @nybList, @n, @len, ret) = {
+        o!(["In getter: ", nybList, n, len]) |
+        // Look up the value of the cell at [map, nybList.slice(0, n + 1)]
+        new valCh in {
+          o!(["Getting from ", nybList.slice(0, n)]) |
+          @[map, nybList.slice(0, n)]!("get", [*valCh]) |
+          for (@val <- valCh) {
+            o!(["Got ", val]) |
+            if (n == len) {
+              o!(["n == len, returning ", val]) |
+              // If we're at the end of the path, return.
+              ret!(val)
+            } else {
+              o!(["nybList.nth(n) = ", nybList.nth(n)]) |
+              // Otherwise check if the rest of the path exists.
+              // Bit k set means cell k exists.
+              // nybList.nth(n) is the cell number
+              // val & powers.nth(nybList.nth(n)) is nonzero if the cell exists
+              // (val/powers.nth(nybList.nth(n))) % 2 is 1 if the cell exists
+              if ((val/powers.nth(nybList.nth(n))) % 2 == 0) {
+                o!(["Cell missing"]) |
+                ret!([])
+              } else {
+                o!(["Cell found, recursing"]) |
+                AsyncMapGetter!(map, nybList, n + 1, len, *ret)
+              }
+            }
+          }
+        }
+      } |
+
+      contract AsyncMap(@"get", @map, @key, ret) = {
+        new hashCh, nybListCh in {
+          // Hash the key to get a 256-bit array
+          @"keccak256Hash"!(key.toByteArray(), *hashCh) |
+          for (@hash <- hashCh) {
+            // Get the bit list
+            ByteArrayToNybbleList!(hash, 0, hash.length(), [], *nybListCh) |
+            for (@nybList <- nybListCh) {
+              AsyncMapGetter!(map, nybList.slice(0, 16), 0, 16, *ret)
```

Comment:
Seems like we're using just (16 nybbles * 4 bits/nybble) = 64 bits of the hash. This is OK for me, but we should probably slice the hash before converting to nybbles, or we waste cycles (and COMMs!)

Reply:
done